modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/PartitionListener.java

```diff
@@ -231,6 +229,10 @@ public void onWrite(Iterator<CommandClosure<WriteCommand>> iterator) {
                 storage.releasePartitionSnapshotsReadLock();
             }
 
+            // Completing the closure out of the partition snapshots lock to reduce possibility of deadlocks as it might
```

Comment:
I didn't get this point. Why it's possible to have a deadlock here?

Reply:
I was able to get into a deadlock with a buggy implementation of `TestPartitionDataStorage` (it was taking an exclusive lock instead of a shared one). But even with this bug fixed, it seems that everything that can be done out of a lock should be done out of it. Especially if it can trigger some code that's hard to imagine right now.